### PR TITLE
[REEF-1683] Use default MaxRetryNumberInRecovery properly

### PR DESCRIPTION
**Pull request by jwang98052**
Currently in IMRUJobDefinitionBuilder, the default _maxRetryNumberInRecovery is set to 0. In REEFIMRUClient and some test code, we bind the value in  IMRUJobDefinition to the named parameter MaxRetryNumberInRecovery. If the user doesn't set this value in IMRUJobDefinitionBuilder when creating IMRUJobDefinition, the default value 0 is bound to the NamedParameter, which override the named parameter MaxRetryNumberInRecovery's own default value.

In IMRUDriver, to avoid this override, we set another default value and replace the value passed in if it 0. That caused some confusion.

This PR makes the change in REEFIMRUClient, if user doesn't pass any value for max retry number, don't bind it to the named parameter.

It also removed the default value for in MaxRetryNumberInRecovery IMRUDriver. Let it use the value either bind by the client or default on the named parameter.

Update the default value to a higher number for our stress testing really need a big number. For functional tests, we already set a proper values in the test code.

JIRA: [REEF-1683](https://issues.apache.org/jira/browse/REEF-1683)
This closes  #

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs [hunk 1]
```diff
@@ -72,7 +72,7 @@ public IMRUJobDefinitionBuilder()
             _updateTaskMemory = 512;
             _coresPerMapper = 1;
             _updateTaskCores = 1;
-            _maxRetryNumberInRecovery = 0;
+            _maxRetryNumberInRecovery = -1;
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: 0 and -1 seem to produce the same behavior in `IMRUDriver` (no retries). But -1 seems weird in context of max number of something. Is there a specific reason to use -1 instead of 0 here?

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/Client/REEFIMRUClient.cs [hunk 2]
```diff
@@ -152,15 +152,18 @@ internal sealed class REEFIMRUClient : IIMRUClient
...
+            if (jobDefinition.MaxRetryNumberInRecovery >= 0)
+            {
+                imruDriverConfiguration.BindNamedParameter(typeof(MaxRetryNumberInRecovery),
+                    jobDefinition.MaxRetryNumberInRecovery.ToString(CultureInfo.InvariantCulture));
+            }
 
             // The JobSubmission contains the Driver configuration as well as the files needed on the Driver.
             var imruJobSubmission = _jobRequestBuilder
-                .AddDriverConfiguration(imruDriverConfiguration)
+                .AddDriverConfiguration(imruDriverConfiguration.Build())
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/Client/REEFIMRUClient.cs]: `imruDriverConfiguration` -> `imruDriverConfigurationBuilder`

**Comment by tcNickolas** [comment]
Is there a way to detect patterns of incorrect use of other parameters in the rest of our code?

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: User can set 0, meaning they don't want to do any retry, or any number greater than 0. -1 is to indicate it is not set by user. Of course user can also set -1, but that doesn't make any sense.

Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: Why don't we keep using 0 as default behavior?

**Comment by jwang98052** [comment]
Usually we only bind named parameter if user set a value. For IMRU, we get config data from IMRUJobDefinition, which contains its own default values. If the user doesn't set those values, its default will override the default in the named parameters. This is a typical case I have seen.  In fact, in IMRUJobDefinitionBuilder, some other variables that have default values also have similar issue and should be revisited.

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: For IMRU, I think the default is to support retry. For IMRUExample and our own functional tests, it is easy to set it from client side. For our big data testing, currently, client use default. Of course, we can ask them to change. But just make our test easy without change anything.

Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: Client can use 0 to specify no retry.

**Comment by jwang98052** [comment]
@tcNickolas I have addressed your comments.

**Comment by jwang98052** [comment]
Another option is to totally respect the default values defined in IMRUBuilderDefinationBuilder. It will override the default values of the named parameters. But it is by its choice. @tcNickolas what do you think?

**Comment by tcNickolas** [comment]
I think that generally a more specific default value (value from some specific workflow) should override a more generic one (the value for parameter). If we're always using either generic default value or user-specified one, why do we need a more specific default value at all?

**Comment by jwang98052** [comment]
Usually we want to set default on named parameter so that user setting becomes optional. 

User can specify their specific default values that override generic default, it is fine. 

The issue is the "user" here is IMRU which is also part of REEF. When default are set at multiple places, that is confusion.

**Comment by tcNickolas** [comment]
Per offline discussion, we should either not have a second (more specific) default value (and rely solely on the generic one) or have two default values and respect more specific one if it is set (the way we do for all other parameters in this code). Since we need second default value for configuration serialization purposes, let's stick with second option, respecting second default value properly. 

We can always use the same value for both defaults (and probably leave a comment for the more specific one saying that its default value should equal the generic default value, so that we're not tempted to modify them out of sync).

File: lang/cs/Org.Apache.REEF.IMRU.Tests/ImruDriverCancelTests.cs [hunk 3]
```diff
@@ -198,8 +198,8 @@ private SystemStateMachine GetDriverState(object driver)
                 .BindNamedParameter(typeof(InvokeGC),
                     jobDefinition.InvokeGarbageCollectorAfterIteration.ToString(CultureInfo.InvariantCulture))
                 .Build();
-                
-            return driverConfig;
+
+            return driverConfigBuilder;
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.IMRU.Tests/ImruDriverCancelTests.cs]: Now that we've reverted to the old way of handling config, we should revert the change in this file. The only thing it does is variable rename, which is not correct anymore (it's `driverConfig` again, since we're building it immediately)

File: lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBrodcastReduceTestBase.cs [hunk 4]
```diff
@@ -134,7 +134,8 @@ public abstract class IMRUBrodcastReduceTestBase : ReefFunctionalTest
             var imruDriverConfiguration = TangFactory.GetTang().NewConfigurationBuilder(new[]
             {
                 driverHandlerConfig,
-                CreateGroupCommunicationConfiguration<TMapInput, TMapOutput, TResult, TPartitionType>(jobDefinition.NumberOfMappers + 1,
+                CreateGroupCommunicationConfiguration<TMapInput, TMapOutput, TResult, TPartitionType>(
```
Comment (tcNickolas) [inline lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBrodcastReduceTestBase.cs]: The change in this file should also be reverted, it's just formatting

**Comment by jwang98052** [comment]
@tcNickolas I have made a clean change and rebased.

**Comment by tcNickolas** [comment]
I'll add a comment to state that new default value is taken from default for named parameter `MaxRetryNumberInRecovery`, test and merge.

We should also note that after this change default IMRU behavior is retry in case of a failure (prior to it retries were disabled by default).
